src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -119,25 +136,68 @@ public function send(string $body, array $headers, int $delay = 0): string
 
-        $this->executeQuery($queryBuilder->getSQL(), [
-            $body,
-            json_encode($headers),
-            $this->configuration['queue_name'],
-            $now,
-            $availableAt,
-        ], [
-            null,
-            null,
-            null,
-            Type::DATETIME,
-            Type::DATETIME,
-        ]);
+        $send = function () use ($queryBuilder, $body, $headers, $now, $availableAt) {
+            $this->executeQuery($queryBuilder->getSQL(), [
+                $body,
+                json_encode($headers),
+                $this->configuration['queue_name'],
+                $now,
+                $availableAt,
+            ], [
+                null,
+                null,
+                null,
+                Type::DATETIME,
+                Type::DATETIME,
+            ]);
+        };
+
+        if ($this->usePgsqlNotify) {
+            if (0 !== $delay) {
+                throw new InvalidArgumentException('Delays are not supported when using PostgreSQL NOTIFY, set the "pgsql_get_notify" option to false to disable this feature. You can also create a dedicated queue for delayed message not using NOTIFY.');
+            }
+
+            $sendAndNotify = function () use ($send) {
+                $this->driverConnection->beginTransaction();
+                $send();
+                $sth = $this->driverConnection->prepare('SELECT pg_notify(?, ?)');
+                $sth->execute([$this->configuration['table_name'], $this->configuration['queue_name']]);
+                $this->driverConnection->commit();
+            };
+
+            try {
+                $sendAndNotify();
+            } catch (TableNotFoundException $e) {
+                $this->driverConnection->rollBack();
+
+                // create table
+                if ($this->autoSetup) {
+                    $this->setup();
+                }
+                $sendAndNotify();
+            }
+        } else {
+            $send();
+        }
 
         return $this->driverConnection->lastInsertId();
     }
 
     public function get(): ?array
     {
+        if ($this->usePgsqlNotify) {
+            if (!$this->listening) {
+                $this->driverConnection->exec('LISTEN "'.$this->configuration['table_name'].'"'); // double quotes MUST be used, so using quote() isn't possible, this shouldn't be an issue because Doctrine doesn't quote table names either.
+                $this->listening = true;
+            // Don't call pgsqlGetNotify() during the first iteration in case messages are already waiting in the queue
+            } else {
+                $notification = $this->driverConnection->getWrappedConnection()->pgsqlGetNotify(\PDO::FETCH_ASSOC, $this->configuration['pgsql_get_notify_timeout']);
+                if (false === $notification || $notification['message'] !== $this->configuration['table_name'] || $notification['payload'] !== $this->configuration['queue_name']) {
+                    return null;
```

Comment:
The table may have contain pending messages even if the worker didn't received any notification. (ie. workers restarts while a message is pushed)

I'd rather use notification for fast delivery but keep the polling to fectch missed messages (with a lower frequency maybe). Fixing it will also allow delayed messages.

Reply:
> The table may contain pending messages even if the worker didn't received any notification.

This case is handled: the `SELECT` query  is always executed during the first call to `get()`.

> Fixing it will also allow delayed messages.

I was think about this, but I don't think it is worth it because it's possible to use another queue for delayed messages.

For delayed message similar to what you suggest, an alternative could be to run the query every N iterations (N being configurable using a dedicated option?). WDYT?

Reply:
> This case is handled: the SELECT query is always executed during the first call to get()
It consume only the first message and the return to a "notification only mode". What is consumer were done for minutes and the tables contains hundred of messsages?

From another side, if you have 100 consumers and the publisher send 100  message/notification the load will be dispatched, each consumer will handle 1 messages, but then the the hunded consumers will try consume next message for nothing.

I really like this idea to add real time on top of the transport, but I think we can improve it. For instance, "reset" the "notification queue" when you are wakedup for "nothing"

Reply:
Actually this is a bug in my current implementation. Postgres sometimes "group" notifications that are the same (same channel, same payload). So anyway we have to consume every message in the queue before waiting. I'll fix this issue. But it doesn't help for the "delay".

Reply:
Bug fixed, we still need to find a way to support delayed messages. What about adding a new option `pg_get_notify_delayed_check`, to force a check for delayed messages every N milliseconds or microseconds?

Reply:
Support for delayed messages added. This basically is what you suggested, but with a configurable delay.